### ISSUE TITLE
[no-use-before-define] class reference in its decorator reports error even if with option `classes: false`

### ISSUE THREAD

**Issue by JounQin** [typescript-eslint/typescript-eslint#2824]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-use-before-define": [
      2,
      {
        "classes": false,
        "functions": false,
      },
    ]
  }
}
```

```TS
@Directive({
  selector: '[rcCidrIpPattern]',
  providers: [
    {
      provide: NG_VALIDATORS,
      useExisting: CidrIpPatternDirective,
      multi: true,
    },
  ],
})
export class CidrIpPatternDirective implements Validator {}
```

**Expected Result**

No error

**Actual Result**

```log
error  'CidrIpPatternDirective' was used before it was defined      @typescript-eslint/no-use-before-define
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.8.2` |
| `@typescript-eslint/parser`        | `4.8.2` |
| `TypeScript`                       | `4.0.5` |
| `ESLint`                           | `7.14.0` |
| `node`                             | `12.19.0` |

**Comment by bradzacher** [comment]
> even if with option `classes: false

you misunderstand - `classes: false` does not allow all references to classes anywhere.
It only allows references to classes in child scopes

```ts
/* eslint no-use-before-define: ['error', { classes: false }] */

CidrIpPatternDirective // banned

function foo() {
  CidrIpPatternDirective // allowed
}

class CidrIpPatternDirective {}
```
[repl](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVzZS1iZWZvcmUtZGVmaW5lOiBbJ2Vycm9yJywgeyBjbGFzc2VzOiBmYWxzZSB9XSAqL1xuXG5DaWRySXBQYXR0ZXJuRGlyZWN0aXZlIC8vIGJhbm5lZFxuXG5mdW5jdGlvbiBmb28oKSB7XG4gIENpZHJJcFBhdHRlcm5EaXJlY3RpdmUgLy8gYWxsb3dlZFxufVxuXG5jbGFzcyBDaWRySXBQYXR0ZXJuRGlyZWN0aXZlIHt9Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMSwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

---

The rule should probably just ignore references in a decorator due to how the transform works
